client/src/pages/Explorer/Explore.js

```diff
@@ -0,0 +1,203 @@
+import Snackbar from "@material-ui/core/Snackbar";
+import Tooltip from "@material-ui/core/Tooltip";
+import MuiAlert from "@material-ui/lab/Alert";
+import ShuffleIcon from "@material-ui/icons/Shuffle";
+
+import useStyles from "../../styles/Explore";
+
+import { useStateContext } from "../../context";
+
+import { getUpdatedList } from "./utils";
+
+function Alert(props) {
+  return <MuiAlert elevation={6} variant="filled" {...props} />;
+}
+
+function FavoriteCheckBox({ place, userId, handleFavoriteChange, openSnack }) {
+  const classes = useStyles();
+  const [checked, setChecked] = useState(place.favorite);
+  return (
+    <>
+      <CustomSmallerCheckBox
+        checked={checked}
+        onChange={(e) => {
+          if (userId) {
+            setChecked(e.target.checked);
+            handleFavoriteChange(e.target.checked, place.name);
+          } else {
+            //@TODO: display login form
+            openSnack();
+          }
+        }}
+        icon={<Favorite className={classes.favoriteDefaultIcon} />}
+        checkedIcon={<Favorite className={classes.favoriteCheckedIcon} />}
+        classes={{ root: classes.customCheckBoxRoot }}
+      />
+    </>
+  );
+}
+
+const Explore = () => {
+  const { user, loading } = useStateContext();
+  const userId = user?._id;
+  const classes = useStyles();
+  const [favorites, setFavorites] = useState([]);
+  const [places, setPlaces] = useState([]);
+  const [snack, setSnack] = useState({ type: "", message: "", open: false });
+
+  const handleShuffleButton = async () => {
+    const returnedArray = await getUpdatedList(places, favorites);
+    setPlaces(returnedArray);
+  };
+
+  const closeSnack = () => {
+    setSnack((prevState) => {
+      return { ...prevState, open: false };
+    });
+  };
+
+  const openSnack = (errorMessage) => {
+    setSnack({
+      type: "info",
+      message: "Please signing or signup first!",
+      open: true,
+    });
+  };
+
+  useEffect(() => {
+    async function getData() {
+      try {
+        if (loading) {
+          return null;
+        } else {
+          let favoriteList = [];
+          //Assuming that the explore page can be accessed by both authenticated and unAuthenticated user.
+          if (userId) {
+            favoriteList = await (
+              await fetch(`/api/users/${userId}/favorite-cities`, {
+                method: "GET",
+                headers: {
+                  "Content-Type": "application/json",
+                },
+              })
+            ).json();
+            setFavorites(favoriteList);
+          }
+          const cityListResponse = await fetch("/api/cities", {
+            method: "GET",
+            headers: {
+              "Content-Type": "application/json",
+            },
+          });
+          const cityList = await cityListResponse.json();
+          cityList.map((city) => {
+            city.favorite = favoriteList.indexOf(city.name) >= 0;
+            return city;
+          });
```

Comment:
technically nested for loop but this is a good example of it doesn't matter because we only have like max 20 destinations